src/esw/fw/bdcmc/Core/Inc/motion_profile.hpp

```diff
@@ -0,0 +1,59 @@
+#include "units/units.hpp"
+
+namespace mrover {
+    template<IsUnit PositionUnit = Radians, IsUnit TimeUnit = Seconds>
+    class TrapezoidalMotionProfile {
+        using VelocityUnit = compound_unit<PositionUnit, inverse<TimeUnit>>;
+        using AccelerationUnit = compound_unit<VelocityUnit, inverse<TimeUnit>>;
+    public:
+        TrapezoidalMotionProfile(PositionUnit initialPosition,
+                      PositionUnit desiredPosition,
+                      VelocityUnit maxVelocity,
+                      AccelerationUnit maxAcceleration) : mMaxVelocity{maxVelocity},
+                                                         mMaxAcceleration{maxAcceleration},
+                                                         mTotalDistance{initialPosition - desiredPosition},
+                                                         mTAccelerationDone{mMaxVelocity / mMaxAcceleration},
+                                                         mTEnd{mTAccelerationDone + mTotalDistance / mMaxVelocity},
+                                                         mTCruiseDone{mTEnd - mTAccelerationDone}
+        {}
+
+        void reset () {
+            mT = 0;
+        }
+
+        void update(TimeUnit dt) {
+            mT += dt;
+        }
+
+        auto velocity() -> VelocityUnit {
```

Comment:
Good stuff dude! Not suggesting you make this change but just for your knowledge: you can actually write code that does this without ever explicitly computing a time schedule for the acceleration changes. The main benefit is that the logic is less cumbersome and the output state is only dependent on the current state meaning you can extend it to use sensor measurements in a "feedback" fashion. 

```c
if ( v^2 / (2 * maxAccel) >= setpointPos - actualPos )  {
   // start slowing down at maxAccel
}
if ( v >= vmax ) {
   // stop accelerating
}
// step kinematics
...
```